### PR TITLE
[Feat][LoginUser and Authentication]- Creation of Login API and AuthN using JWT

### PR DESCRIPTION
**Pull request by TreshAnn**
## Ticket/s
- [Authentication](https://trello.com/c/SGKkX1mB/3-authentication)
- [API- LoginUser](https://trello.com/c/kH6fbmJ7/7-api-login)

### LoginUser
- Add feature loginUser API
- Add UnAuthenticatedError
- Add routing for loginUser
- Add function for comparePassword
- Modify interface name

### Authentication
- Add JWT function for token generation

## Notes
- installed morgan dependency for logging request details
- installed jsonwebtoken for User Authentication

#### via Postman cURL

- Open Postman
- click import tab, paste cURL
##### Note: Email and Password must be existing in DB 
```
curl --location 'http://localhost:5000/api/v1/auth/login' \
--header 'Content-Type: application/json' \
--data-raw '{
    "email": "tresh@gmail.com",
    "password": "password"
}'
```
#### Validate Token
- [Validate Token Here](https://jwt.io/)
- copy token from postman and paste it here

## Expected Test Result
-  if request body has correct  email, and password, it will send back 201 OK status code and user object with generated token

image

- if request body has incorrect email and password, it will send back 401 Unauthorized status code and error message

image

- if email and password is missing from request body

image

### CODE REVIEW

File: backend/src/models/User.interface.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-export interface IRegisterUserRq{
+export interface IUserRq{
```
Comment (dandalandone) [inline backend/src/models/User.interface.ts]: separate interface of login and user register

File: backend/src/server.ts [hunk 2]
```diff
@@ -14,6 +15,10 @@ import authRouter from './routes/authRouter'
 import notFoundMiddleware from './middleware/not-found'
 import errorHandlerMiddleware from './middleware/error-handler'
 
+if(process.env.NODE_ENV !== 'production'){
+    app.use(morgan('dev'))
+}
+
```
Comment (dandalandone) [inline backend/src/server.ts]: remove

File: backend/src/server.ts [hunk 3]
```diff
@@ -3,6 +3,7 @@ import express from 'express'
 const app = express()
 import dotenv from 'dotenv'
 dotenv.config()
+import morgan from 'morgan'
```
Comment (dandalandone) [inline backend/src/server.ts]: remove

File: backend/package.json [hunk 4]
```diff
@@ -19,16 +20,20 @@
   },
   "dependencies": {
     "@types/bcryptjs": "^2.4.2",
+    "@types/jsonwebtoken": "^9.0.2",
```
Comment (dandalandone) [inline backend/package.json:23]: remove

File: backend/package.json [hunk 5]
```diff
@@ -19,16 +20,20 @@
...
     "bcryptjs": "^2.4.3",
+    "body-parser": "^1.20.2",
     "dotenv": "^16.3.1",
     "express": "^4.18.2",
     "express-async-errors": "^3.1.1",
     "http-status-codes": "^2.2.0",
+    "jsonwebtoken": "^9.0.1",
     "mongodb": "^5.6.0",
     "mongoose": "^7.3.1",
+    "morgan": "^1.10.0",
```
Comment (dandalandone) [inline backend/package.json]: remove

**Review by dandalandone — CHANGES_REQUESTED** [review]
Please check my comments, thank you!

File: backend/src/models/User.interface.ts (hunk 1 above)
Comment (TreshAnn) [reply backend/src/models/User.interface.ts]: Done , Sir!

Comment (DanDalanDone69) [reply backend/src/models/User.interface.ts]: Change to ILoginRq

Comment (DanDalanDone69) [reply backend/src/models/User.interface.ts]: Change to ILoginRq

**Review by dandalandone — APPROVED** [review]
LGTM!

**Pull request merged**